lib/server-builder.js

```diff
@@ -0,0 +1,161 @@
+    await fs.writeFile(propertiesPath, newProperties, 'utf8');
+  }
+
+  updateGradleDistUrl (propertiesContent, version) {
+    return propertiesContent.split('\n')
+      .map(function (line) {
+        if (line.startsWith(GRADLE_URL_PREFIX)) {
+          return `${GRADLE_URL_PREFIX}${GRADLE_URL_TEMPLATE.replace('VERSION', version)}`;
+        } else {
+          return line;
+        }
+      })
+      .join('\n');
+  }
+
+  async insertAdditionalAppDependencies (additionalAppDependencies) {
+    const buildPath = path.resolve(this.serverPath, 'app', 'build.gradle');
+    const originalConfiguration = await fs.readFile(buildPath, 'utf8');
+    const newConfiguration = this.updateDependencyLines(originalConfiguration, additionalAppDependencies);
+    await fs.writeFile(buildPath, newConfiguration, 'utf8');
+  }
+
+  updateDependencyLines (configurationContent, additionalAppDependencies) {
+    if (!Array.isArray(additionalAppDependencies)) {
+      throw new Error('additionalAppDependencies must be an array');
+    }
+    const dependencyLines = additionalAppDependencies
+      .map(function (dependency) {
+        if (/[\s']/.test(dependency)) {
+          throw new Error(`Disallowed characters in an additional dependency: ${dependency}`);
+        }
+        return `implementation '${dependency}'`;
+      })
+      .join('\n');
+
+    return configurationContent.split('\n')
+      .map(function (line) {
+        if (line.trim() === GRADLE_DEPENDENCIES_PLACEHOLDER) {
+          return dependencyLines;
+        } else {
+          return line;
+        }
+      })
+      .join('\n');
+  }
+
+  async runBuildProcess () {
+    const {cmd, args} = this.getCommand();
+    log.debug(`Beginning build with command '${cmd} ${args.join(' ')}' ` +
+      `in directory '${this.serverPath}'`);
+    const gradlebuild = new SubProcess(cmd, args, {
+      cwd: this.serverPath,
+      stdio: ['ignore', 'pipe', 'pipe'],
+    });
+    let buildLastLines = [];
+
+    const logMsg = `Output from Gradle ${this.showGradleLog ? 'will' : 'will not'} be logged`;
+    log.debug(`${logMsg}. To change this, use 'showGradleLog' desired capability`);
+    gradlebuild.on('stream-line', line => {
+      // if we have an error we want to output the logs
+      // otherwise the failure is inscrutible
+      // but do not log expected errors
+      const ignoreError = IGNORED_ERRORS.some((x) => line.includes(x));
+
+      if (this.showGradleLog && !ignoreError) {
+        gradleLog.error(line);
+      }
+      buildLastLines.push(`${EOL}${line}`);
+      if (buildLastLines.length > GRADLE_MAX_ERROR_LOG_LINES) {
+        buildLastLines = buildLastLines.slice(-GRADLE_MAX_ERROR_LOG_LINES);
+      }
+    });
+
+    buildLastLines = [];
+
+    try {
+      await gradlebuild.start();
+      await gradlebuild.join();
+    } catch (err) {
+      let msg = `Unable to build Espresso server - ${err}\n` +
+        `Gradle error message:${EOL}${buildLastLines}`;
```

Comment:
should we really collect these lines if we have `showGradleLog`? I would just suggest setting it to true if it is not set and checking the server log.

Reply:
That will complicate debugging flaky issues with building. So I would still like to have output for build failures, at least by default.
Also I think XCUITest driver prints error log when similar capability is `false`.